[DropdownMenu][Popover][Dialog] Apply explicit trigger auto focus

andy-hook opened the pull request:
This change applies the focus fixes first added in https://github.com/radix-ui/primitives/pull/739 but now accounts for the two modality modes.
The above changes were intentionally removed during a rebase to make their addition easier to review and cleaner in history.

andy-hook commented on packages/react/dialog/src/Dialog.tsx:
```diff
@@ -347,6 +339,19 @@ const DialogContentImpl = React.forwardRef((props, forwardedRef) => {
+              const originalEvent = event.detail.originalEvent;
+              const ctrlLeftClick = originalEvent.button === 0 && originalEvent.ctrlKey === true;
+              const isRightClick = originalEvent.button === 2 || ctrlLeftClick;
+              preventAutoFocusRef.current = isRightClick || !props.disableOutsidePointerEvents;
```
Rather than focusing on a left pointer down I've instead excluded the auto focus when right click or pointer events enabled (e.g. non modal).

This was the cleanest way I could see to handle the differing behaviour between the modality modes, let me know what you think though.

jjenzz replied:
I feel like I must be missing something here because my immediate thought is to wonder why we can't have the separate implementations in the different modal type components... like:

```js
const DialogContentModal = () => {
	// ...
	return (
		<DialogContentImpl 
			disableOutsidePointerEvents
			onUnmountAutoFocus={composeEventHandlers(onCloseAutoFocus, (event) => {
          		event.preventDefault();
          		if (!preventAutoFocusRef.current) context.triggerRef.current?.focus();
        	})}
			onPointerDownOutside={event => {
				const originalEvent = event.detail.originalEvent;
				const ctrlLeftClick = originalEvent.button === 0 && originalEvent.ctrlKey === true;
				const isRightClick = originalEvent.button === 2 || ctrlLeftClick;
				preventAutoFocusRef.current = isRightClick
			}}
		/>
	);
}

// Doesn't need the trigger focus because pointer events aren't disabled?
const DialogContentNonModal = () => {
	// ...
	return (
		<DialogContentImpl 
			disableOutsidePointerEvents
			onUnmountAutoFocus={composeEventHandlers(onCloseAutoFocus, (event) => {
				event.preventDefault();
			})}
		/>
	);
}

jjenzz replied:
Although, I'm curious why trigger focus doesn't happen when `disableOutsidePointerEvents` is `false` in your logic here? I think the original ticket was to specifically explicitly focus the trigger but this looks like it wouldn't get explicitly focused now in a non-modal dialog?

andy-hook replied:
Yeah we can totally do this though I stopped short because we can't do it for `DropdownMenu` unless we move the trigger to `Menu`. I suppose we could have the implementation be different for that case.. or even bite the bullet and move the trigger?

andy-hook commented on packages/react/dropdown-menu/src/DropdownMenu.tsx:
```diff
@@ -225,8 +228,21 @@ const DropdownMenuRootContent = React.forwardRef((props, forwardedRef) => {
+          const originalEvent = event.detail.originalEvent;
+          const ctrlLeftClick = originalEvent.button === 0 && originalEvent.ctrlKey === true;
+          const isRightClick = originalEvent.button === 2 || ctrlLeftClick;
+          preventAutoFocusRef.current = isRightClick || !context.modal;
```
I don't like that we need to pass `modal` down in context to make this work. I've had to do this because we have no access to the trigger at the `Menu` level. An alternative is to move the trigger into `Menu` but I'm a little concerned at the sheer number of concerns that `Menu` is already handling.

If anyone can think of an alternative I'd be open to it.

benoitgrelard replied:
It doesn't bother me too much, I think ultimately this is easier to understand than if we had to move around the whole trigger requirement into `Menu`. We've tried that before and it was pretty confusing if I remember correctly.

andy-hook replied:
Yeah true, I really didn't really want to go there a third time haha

andy-hook commented on packages/react/menu/src/Menu.tsx:
```diff
@@ -301,23 +300,9 @@ const MenuRootContentNonModal = React.forwardRef((props, forwardedRef) => {
-        },
+      onCloseAutoFocus={composeEventHandlers(
+        props.onCloseAutoFocus,
+        (event) => event.preventDefault(),
```
`Menu` will now always expect auto focus will be managed manually (where before it was leaning into the implicit auto focus behaviour hence it working fine with dropdown in the previous config)

benoitgrelard commented:
> Focus trigger on any pointerDown outside unless it's a "right click"

Did you consider the overlay too? I think we were saying from modal right click is allowed but only if there's no overlay right? ie. all but Dialog.

andy-hook commented:
> Did you consider the overlay too? I think we were saying from modal right click is allowed but only if there's no overlay right? ie. all but Dialog.

I was referring to the auto focus of the trigger which I understood as never wanting to do that on right click? or were you referring to the "focusing of another element on right click" thing? e.g. I can right click on a textarea to copy some text even when modal?

benoitgrelard commented:
Yeah I was referring to the latter. Sorry about that, so you mean this PR is purely only about the former?

benoitgrelard commented:
Although it's kinda related, because when right clicking outside of a dialog which has an overlay, I would expect the focus to go back to the trigger.

andy-hook commented:
> Yeah I was referring to the latter. Sorry about that, so you mean this PR is purely only about the former?

It was for the former at the time as the latter relies on [this change ](https://github.com/radix-ui/primitives/pull/772) but I can probably update this now if needed.

> Although it's kinda related, because when right clicking outside of a dialog which has an overlay, I would expect the focus to go back to the trigger.

Maybe my wires are crossed but I assumed we didn't want anything to happen when right clicking an overlay? e.g. it would remain open so focusing the trigger then wouldn't make sense?

benoitgrelard commented:
> Maybe my wires are crossed but I assumed we didn't want anything to happen when right clicking an overlay? e.g. it would remain open so focusing the trigger then wouldn't make sense?

My bad, I'm the one who's getting confused, you're 100% correct.
This is hurting my brain… 🤯

andy-hook commented:
@benoitgrelard @jjenzz – I've updated the PR with the suggested change

> Yeah I was referring to the latter.

On this point, I updated the branch with [this change](https://github.com/radix-ui/primitives/pull/772) which seems to get us most of the way there (@benoitgrelard you win, I actually thought this wouldn't be possible at all!)

https://user-images.githubusercontent.com/11708259/125306496-2fcb1480-e327-11eb-815a-fef9e6b4e39d.mp4

benoitgrelard commented:
Amazing! Yeah that's what I thought that this is what would fix it. So good 🙂
Is this ready for review now then?

andy-hook commented:
> Is this ready for review now then?

👍

benoitgrelard commented on packages/react/dialog/src/Dialog.tsx:
```diff
@@ -221,15 +222,19 @@ const DialogContentModal = React.forwardRef((props, forwardedRef) => {
-              originalEvent.button === 2 ||
-              (originalEvent.button === 0 && originalEvent.ctrlKey === true);
+            const originalEvent = event.detail.originalEvent;
+            const ctrlLeftClick = originalEvent.button === 0 && originalEvent.ctrlKey === true;
```
I like this! `ctrlLeftClick` 👍

jjenzz replied:
agree, i liked this clarity too :+1:

benoitgrelard commented on packages/react/dialog/src/Dialog.tsx:
```diff
@@ -221,15 +222,19 @@ const DialogContentModal = React.forwardRef((props, forwardedRef) => {
           disableOutsidePointerEvents
+          onCloseAutoFocus={composeEventHandlers(props.onCloseAutoFocus, (event) => {
+            event.preventDefault();
+            if (!isRightClickOutsideRef.current) context.triggerRef.current?.focus();
```
Do we actually need this check here?

Because I see below that we say we don't close if it's a right click, therefore I'm thinking we never get to `onCloseAutoFocus` in the case of a right click?

benoitgrelard commented on packages/react/dialog/src/Dialog.tsx:
```diff
@@ -244,24 +249,20 @@ const DialogContentModal = React.forwardRef((props, forwardedRef) => {
 const DialogContentNonModal = React.forwardRef((props, forwardedRef) => {
   const context = useDialogContext(CONTENT_NAME);
-  const isPointerDownOutsideRef = React.useRef(false);
+  const isEscapeKeyPressRef = React.useRef(false);
```
nit: Shall we name this `isEscapeKeyDown` considering that's the name of the handler?

benoitgrelard commented on packages/react/popover/src/Popover.tsx:
```diff
@@ -224,6 +225,21 @@ const PopoverContentModal = React.forwardRef((props, forwardedRef) => {
           // (closed !== unmounted when animating out)
           trapFocus={context.open}
           disableOutsidePointerEvents
+          onCloseAutoFocus={composeEventHandlers(props.onCloseAutoFocus, (event) => {
+            event.preventDefault();
+            if (!isRightClickOutsideRef.current) context.triggerRef.current?.focus();
+          })}
+          onPointerDownOutside={composeEventHandlers(
+            props.onPointerDownOutside,
+            (event) => {
+              const originalEvent = event.detail.originalEvent;
+              const ctrlLeftClick = originalEvent.button === 0 && originalEvent.ctrlKey === true;
+              const isRightClick = originalEvent.button === 2 || ctrlLeftClick;
+
+              isRightClickOutsideRef.current = isRightClick;
+            },
+            { checkForDefaultPrevented: false }
+          )}
```
👍

benoitgrelard commented on packages/react/popover/src/Popover.tsx:
```diff
@@ -240,7 +256,7 @@ const PopoverContentModal = React.forwardRef((props, forwardedRef) => {
   const { portalled = true, ...contentNonModalProps } = props;
   const context = usePopoverContext(CONTENT_NAME);
-  const isPointerDownOutsideRef = React.useRef(false);
+  const isEscapeKeyPressRef = React.useRef(false);
```
Same nit on the name here

benoitgrelard commented on packages/react/dropdown-menu/src/DropdownMenu.tsx:
```diff
@@ -215,6 +217,7 @@ type DropdownMenuRootContentPrimitive = Polymorphic.ForwardRefComponent<
 const DropdownMenuRootContent = React.forwardRef((props, forwardedRef) => {
   const { portalled = true, ...contentProps } = props;
   const context = useDropdownMenuContext(CONTENT_NAME);
+  const preventAutoFocusRef = React.useRef(false);
```
nit: can we rename this more explicitly `preventCloseAutoFocusRef`?

benoitgrelard reviewed (commented):
It looks good to me overall, just left one question and a few nits.

jjenzz commented on packages/react/dropdown-menu/src/DropdownMenu.tsx:
```diff
@@ -225,8 +228,21 @@ const DropdownMenuRootContent = React.forwardRef((props, forwardedRef) => {
+      })}
+      onEscapeKeyDown={composeEventHandlers(props.onEscapeKeyDown, () => {
+        preventCloseAutoFocusRef.current = false;
       })}
```
Do we need `{ checkForDefaultPrevented: false }` here? In fact, do we need it on all of the handlers that set the handlers in this PR that set `isPointerDownOutsideRef` or `preventCloseAutoFocusRef` ?

benoitgrelard replied:
I thought we said we would re-assess all this separately?

andy-hook replied:
Honestly I wasn't on which way to go here, we did say we'd assess it separately so I just carried over the existing stance from the pointer handlers in `Popover` and `Menu`.

andy-hook replied:
Oh that's true, i'll revert it on `onEscapeKeyDown` just to keep it inline with how it was for now.
